read-fonts/src/collections/range_set.rs

```diff
@@ -0,0 +1,395 @@
+            }
+        };
+
+        // There may be one or more ranges proceeding this new range that intersect, find and merge them as needed.
+        loop {
+            let Some((next_start, next_end)) = self.next_range(start) else {
+                // No existing ranges which might overlap, can now insert the current range
+                self.ranges.insert(start, end);
+                return;
+            };
+
+            if range_is_subset(start, end, next_start, next_end) {
+                return;
+            }
+            if ranges_overlap_or_adjacent(start, end, next_start, next_end) {
+                start = min(start, next_start);
+                end = max(end, next_end);
+                self.ranges.remove(&next_start);
+            } else {
+                self.ranges.insert(start, end);
+                return;
+            }
+        }
+    }
+
+    /// Returns an iterator over the contained ranges.
+    pub fn iter(&'_ self) -> impl Iterator<Item = RangeInclusive<T>> + '_ {
+        self.ranges.iter().map(|(a, b)| *a..=*b)
+    }
+
+    /// Returns an iterator over the intersection of this and other.
+    pub fn intersection<'a>(
+        &'a self,
+        other: &'a Self,
+    ) -> impl Iterator<Item = RangeInclusive<T>> + 'a {
+        IntersectionIter {
+            it_a: self.iter().peekable(),
+            it_b: other.iter().peekable(),
+        }
+    }
+
+    /// Finds a range in this set with a start greater than or equal to the provided start value.
+    fn next_range(&self, start: T) -> Option<(T, T)> {
+        let (next_start, next_end) = self.ranges.range(start..).next()?;
+        Some((*next_start, *next_end))
+    }
+
+    /// Finds a range in this set with a start less than the provided start value.
+    fn prev_range(&self, start: T) -> Option<(T, T)> {
+        let (next_start, next_end) = self.ranges.range(..start).next_back()?;
+        Some((*next_start, *next_end))
+    }
+}
+
+impl<T> Extend<RangeInclusive<T>> for RangeSet<T>
+where
+    T: Copy + Ord + OrdAdjacency<T>,
+{
+    fn extend<I: IntoIterator<Item = RangeInclusive<T>>>(&mut self, iter: I) {
+        iter.into_iter().for_each(|r| self.insert(r));
+    }
+}
+
+impl<T> FromIterator<RangeInclusive<T>> for RangeSet<T>
+where
+    T: Default + Copy + Ord + OrdAdjacency<T>,
+{
+    fn from_iter<I: IntoIterator<Item = RangeInclusive<T>>>(iter: I) -> Self {
+        let mut result: Self = Default::default();
+        result.extend(iter);
+        result
+    }
+}
+
+impl<T> Debug for RangeSet<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(f, "RangeSet {{")?;
+        for (start, end) in self.ranges.iter() {
+            write!(f, "[{:?}, {:?}], ", start, end)?;
+        }
+        write!(f, "}}")
+    }
+}
+
+struct IntersectionIter<A, B, T>
+where
+    A: Iterator<Item = RangeInclusive<T>>,
+    B: Iterator<Item = RangeInclusive<T>>,
```

Comment:
we can delete the bounds here I think, as long as they're on the impl.

Reply:
Peekable requires the types to be iterators.